iModelCore/ecobjects/src/SchemaMerger.cpp

```diff
@@ -218,8 +236,30 @@ BentleyStatus SchemaMerger::MergeSchemas(SchemaMergeResult& result, bvector<ECSc
     //Make a copy of the input vectors so we don't modify the original (given to us as const anyways)
     bvector<ECSchemaCP> left(rawLeft);
     bvector<ECSchemaCP> right(rawRight);
-    ECSchema::SortSchemasInDependencyOrder(left);
-    ECSchema::SortSchemasInDependencyOrder(right);
+    auto doNotMergeReferences = options.DoNotMergeReferences();
+    if(!doNotMergeReferences)
+        {
+        ECSchema::SortSchemasInDependencyOrder(left);
+        ECSchema::SortSchemasInDependencyOrder(right);
+        }
+    else
+        {
+        bool isValid = true;
+        if(SchemaMerger::ValidateUniqueSchemaNames(left) != BentleyStatus::SUCCESS)
+            {
+            result.Issues().ReportV(IssueSeverity::Fatal, IssueCategory::BusinessProperties, IssueType::ECSchema, ECIssueId::EC_0060,
+                "At least a single existing schema name was non-unique.");

```

Comment:
This can probably be improved, the message as it stands will not be very useful for troubleshooting. I'd say make it log the side (left or right) and also make it mention the name of the schema which was non unique. (ValidateUniqueSchemaNames could return an optional string, or it could directly log any duplicate schema names if there are multiple) That way the callers can much more easily track the issue down.
Then you do not even need the isValid flag, just call Validate for both sides in a single line, the method itself logs any violations and if one of the two calls fails, you return error here.